### PR TITLE
Adding test coverage for SupportedOSPlatform on default windows values

### LINKED ISSUE #13494: SupportedOSPlatform assembly attributes are not getting generated for version less platform name
### SupportedOSPlatform assembly attributes are not getting generated for version less platform name

In https://github.com/dotnet/runtime/issues/39989#issuecomment-665197220 we decided to rename MinimumOSPlatformAttribute into RemovedInOSPlatformAttribute and support version less platform name.  We had renamed the attribute with https://github.com/dotnet/sdk/pull/12775 but the version less attribute name support is missing

 ```xml
<PropertyGroup>
   <TargetFramework>net5.0-windows</TargetFramework>
</PropertyGroup>
```
Only works when adding a valid version
```xml
<PropertyGroup>
   <TargetFramework>net5.0-windows7.0</TargetFramework>
</PropertyGroup>
```
CC @adamsitnik @jeffhandley @terrajobst @wli3

### PR DESCRIPTION
**Pull request by sfoslund**
Fixes #13494

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Windows.targets [hunk 1]
```diff
@@ -27,6 +27,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   <PropertyGroup Condition="'$(TargetPlatformIdentifier)' == 'windows' and '$(TargetPlatformVersion)' == ''" >
     <TargetPlatformVersion>7.0</TargetPlatformVersion>
     <!-- Automatic generation of TargetPlatformMoniker and TargetPlatformDisplayName have already been evaluated so we need to hardcode these values. -->
+    <SupportedOSPlatform>7.0</SupportedOSPlatform>
     <TargetPlatformMoniker>Windows,Version=7.0</TargetPlatformMoniker>
     <TargetPlatformDisplayName>Windows 7.0</TargetPlatformDisplayName>
```
Comment (terrajobst) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Windows.targets:32]: Any reason we don't use `$(TargetPlatformVersion)` that we defined right above? Seems odd to duplicate the version number.

```suggestion
    <SupportedOSPlatform>$(TargetPlatformVersion)</SupportedOSPlatform>
    <TargetPlatformMoniker>Windows,Version=$(TargetPlatformVersion)</TargetPlatformMoniker>
    <TargetPlatformDisplayName>Windows $(TargetPlatformVersion)</TargetPlatformDisplayName>
```

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Windows.targets [hunk 2]
```diff
@@ -27,6 +27,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   <PropertyGroup Condition="'$(TargetPlatformIdentifier)' == 'windows' and '$(TargetPlatformVersion)' == ''" >
     <TargetPlatformVersion>7.0</TargetPlatformVersion>
     <!-- Automatic generation of TargetPlatformMoniker and TargetPlatformDisplayName have already been evaluated so we need to hardcode these values. -->
+    <SupportedOSPlatform>7.0</SupportedOSPlatform>
```
Comment (terrajobst) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Windows.targets:30]: Is that the right name? I thought [we agreed](https://github.com/dotnet/designs/pull/157) on `SupportedOSPlatformVersion`.

/cc @dsplaisted

Comment (sfoslund) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Windows.targets:30]: It looks like it got renamed to `SupportedOSPlatform` [here](https://github.com/dotnet/sdk/pull/12775/files#diff-b1cca2205c228925d515de229057c7c6R81).

**Comment by jeffhandley** [comment]
I was anticipating this PR would make changes to [this block of code](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L119-L126) in `Microsoft.NET.GenerateAssemblyInfo.targets`:

```xml
    <ItemGroup Condition="'$(TargetPlatformIdentifier)' != ''
                         and '$(SupportedOSPlatform)' != ''
                         and '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
                         and $([MSBuild]::VersionGreaterThanOrEquals($(_TargetFrameworkVersionWithoutV), '5.0'))">
      <AssemblyAttribute Include="System.Runtime.Versioning.SupportedOSPlatformAttribute">
        <_Parameter1>$(TargetPlatformIdentifier)$(SupportedOSPlatform)</_Parameter1>
      </AssemblyAttribute>
    </ItemGroup>
```

The way I interpret that, we won't emit the `SupportedOSPlatformAttribute` unless there is a version specified on the platform (held by `$(SupportedOSPlatform)`.  @terrajobst is it correct that:

1. `net5.0-windows` should result in `[SupportedOSPlatformAttribute("windows")]`, and
2. `net5.0-ios13.0` should result in `[SupportedOSPlatformAttribute("ios13.0")]`

**Comment by wli3** [comment]
I'll defer to @dsplaisted and PMs on this topic. The code change is the small part. It is more like a spec change of how to derive these properties.

**Comment by sfoslund** [comment]
> The way I interpret that, we won't emit the SupportedOSPlatformAttribute unless there is a version specified on the platform (held by $(SupportedOSPlatform)

@jeffhandley I see, I misunderstood the ask. So we only want the attribute to include the version when the TargetPlatformVersion was specified explicitly by the user? That is, we won't include it when we're using the default target platform version?

**Comment by jeffhandley** [comment]
> So we only want the attribute to include the version when the TargetPlatformVersion was specified explicitly by the user? That is, we won't include it when we're using the default target platform version?

That was my understanding.  As right now, specifying `net5.0-windows` does not produce the assembly-level attribute at all because the `ItemGroup` has a condition requiring a non-empty version.  But we should wait for @terrajobst to weigh in authoritatively.

**Comment by buyaa-n** [comment]
> @jeffhandley I see, I misunderstood the ask. So we only want the attribute to include the version when the TargetPlatformVersion was specified explicitly by the user? That is, we won't include it when we're using the default target platform version?

Is the existing logic producing the default version when the version is not provided by the user? Which means `net5.0-windows` would produce `[SupportedOSPlatformAttribute("windows7.0")]`? Well, that would be wrong as for analyzer version less means 0.0 not 7.0. `net5.0-windows` should produce `[SupportedOSPlatformAttribute("windows")]`
`

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibraryWithOSSupportedVersion.cs [hunk 3]
```diff
@@ -72,6 +72,21 @@ public void WhenSupportedOSPlatformISNotSetTargetPlatformVersionIsSetItCanGenera
...
+            TestProject testProject = SetUpProject();
+            testProject.AdditionalProperties["TargetPlatformIdentifier"] = "windows";
+
+            var testAsset = _testAssetsManager.CreateTestProject(testProject);
+
+            var runCommand = new DotnetCommand(Log, "run");
+            runCommand.WorkingDirectory = Path.Combine(testAsset.TestRoot, testProject.Name);
+            runCommand.Execute()
+                .Should()
+                .Pass().And.HaveStdOutContaining("PlatformName:Windows7.0");
```
Comment (buyaa-n) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibraryWithOSSupportedVersion.cs:87]: From this test seems it is producing `7.0` by default

**Comment by jeffhandley** [comment]
It sounds like we might need some acceptance criteria laid out.  Given each of the following TFMs, what attributes should get produced, @terrajobst?

1. `net5.0`
2. `net5.0-windows`
3. `net5.0-windows7.0`
4. `net5.0-windows10.0.19041`

Note that `net5.0-windows10` currently yields a build error that `10.0.0.0 is not a valid TargetPlatformVersion for Windows`, which I presume to be expected.

**Comment by sfoslund** [comment]
> Is the existing logic producing the default version when the version is not provided by the user?

Yes, the target platform version property needs to be populated so we supply the default of 7.0 for windows when the user doesn't specify one

**Comment by buyaa-n** [comment]
> the target platform version property needs to be populated so we supply the default of 7.0 for windows when the user doesn't specify one

Thanks, @sfoslund for clarification, is there any way to keep it unspecified? if not could we at least default to `0.0`.? If the default is `7.0` for `windows` what about other platforms?

**Comment by jeffhandley** [comment]
I think it's OK to default to 7.0 for Windows. Our attributes are applied to APIs as `[SupportedOSPlatform("windows")]`, but having the project gain `[SupportedOSPlatform("windows7.0")]` will mean the project has a version greater than 0.0, which keeps things working nicely.  But we're still relying on being able to have `net5.0-windows` result in an attribute being produced (even if it has the default version of 7.0).

I think the answers to the acceptance criteria list above will be:

1. `net5.0` ==> No attributes
2. `net5.0-windows` ==> `[SupportedOSPlatform("Windows7.0")]`
3. `net5.0-windows7.0` ==> `[SupportedOSPlatform("Windows7.0")]`
4. `net5.0-windows10.0.19041` ==> `[SupportedOSPlatform("Windows10.0.19041")]`

But I know @terrajobst was going to take a look and weigh in for the authoritative response.

**Comment by buyaa-n** [comment]
> I think it's OK to default to 7.0 for Windows. Our attributes are applied to APIs as `[SupportedOSPlatform("windows")]`, but having the project gain `[SupportedOSPlatform("windows7.0")]` will mean the project has a version greater than 0.0, which keeps things working nicely. But we're still relying on being able to have `net5.0-windows` result in an attribute being produced (even if it has the default version of 7.0).

That is true, seems fine for APIs applied `[UnsupportedOSPlatform("windows")]` attribute too. @sfoslund never mind 😄

**Comment by terrajobst** [comment]
@jeffhandley @sfoslund 

> > So we only want the attribute to include the version when the TargetPlatformVersion was specified explicitly by the user? That is, we won't include it when we're using the default target platform version?
> 
> That was my understanding. As right now, specifying `net5.0-windows` does not produce the assembly-level attribute at all because the `ItemGroup` has a condition requiring a non-empty version. But we should wait for @terrajobst to weigh in authoritatively.

Not quite. We want development assets to stay clean of the version number, but no durable artifacts. Otherwise, the logic what the implicit version number was has to be duplicated in multiple places. We want the SDK to be authority. Hence, the assembly attributes (and NuGet packages) should contain the actual version number, regardless of whether the user explicitly set or whether it was implicitly set by the SDK.

Does this make sense?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Windows.targets (hunk 2 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Windows.targets]: We decided to rename it to SupportedOSPlatformVersion, but haven't done so yet.  I was planning to do that as part of #13453

**Comment by sfoslund** [comment]
> This raises the question how a developer could control this. One option is to say the developer explicitly sets <SupportedOSPlatformVersion> to 0.0. In this case, the generated attribute should not have a version number.

Yes, that makes sense. If that's the schema we want to go with I can implement it in this PR.

**Comment by jeffhandley** [comment]
@sfoslund / @terrajobst Does that get us to the point of needing to make the following changes?

1. When a version-less platform identifier is used (`net5.0-windows`), we will:
    * Apply the default version (in this case, `7.0`)
    * Emit the attribute (in this case `[SupportedOSPlatform("Windows7.0")]`)
2. Support a `<SupportedOSPlatformVersion>` property to override that default
    * This allows `0.0` to be specified to result in `[SupportedOSPlatform("Windows")]`

**Comment by terrajobst** [comment]
Perfect!

**Comment by jeffhandley** [comment]
Cool; thanks.  Then we have acceptance criteria of:

1. `net5.0` ==> No attributes
2. `net5.0-windows` ==> `[SupportedOSPlatform("Windows7.0")]`
3. `net5.0-windows` + `<SupportedOSPlatformVersion>0.0</SupportedOSPlatformVerison>` ==> `[SupportedOSPlatform("Windows")]`
4. `net5.0-windows7.0` ==> `[SupportedOSPlatform("Windows7.0")]`
5. `net5.0-windows10.0.19041` ==> `[SupportedOSPlatform("Windows10.0.19041")]`

**Comment by terrajobst** [comment]
> 5\. `net5.0-windows10.0.19041` ==> `[SupportedOSPlatform("Windows10.0.19041")]`

Assuming the .NET 5 SDK supports that version of Windows.

**Comment by sfoslund** [comment]
> >5. net5.0-windows10.0.19041 ==> [SupportedOSPlatform("Windows10.0.19041")]

>I haven't checked, but I assume the .NET 5 SDK supports that version of Windows?

The versions of windows the SDK supports can be found here: https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.WindowsSdkSupportedTargetPlatforms.props
Note that the 10.X versions have 4 parts.

**Comment by sfoslund** [comment]
@terrajobst I've updated to the criteria from [this comment](https://github.com/dotnet/sdk/pull/13507#issuecomment-692405455). Look okay to you?

**Review by dsplaisted — APPROVED** [review]
Looks good, but it looks like the tests aren't covering the below cases via setting the `TargetFramework` as below.  Rather, the `TargetFramework` for the tests always seems to be `net5.0`, and then `TargetPlatformIdentifier` and `TargetPlatformVersion` are set as appropriate.

It would be nice to test actually setting the `TargetFramework` as specified below, as that is the more common scenario (and uses a different code path to set some of the properties if 7.0 is inferred as the platform version).

> 1. `net5.0` ==> No attributes
> 2. `net5.0-windows` ==> `[SupportedOSPlatform("Windows7.0")]`
> 3. `net5.0-windows` + `<SupportedOSPlatformVersion>0.0</SupportedOSPlatformVerison>` ==> `[SupportedOSPlatform("Windows")]`
> 4. `net5.0-windows7.0` ==> `[SupportedOSPlatform("Windows7.0")]`
> 5. `net5.0-windows10.0.19041` ==> `[SupportedOSPlatform("Windows10.0.19041")]`

**Review by jeffhandley — APPROVED** [review]
It looks like this is still using `<SupportedOSPlatform>` to indicate the version instead of `<SupportedOSPlatformVersion>` as was mentioned in the latest comments.

I don't know if that can or should be changed with this PR or separately though.

**Comment by sfoslund** [comment]
> It looks like this is still using SupportedOSPlatform to indicate the version instead of SupportedOSPlatformVersion as was mentioned in the latest comments.

I believe @dsplaisted is changing that in a separate PR.

**Pull request merged**